staker/reward_math.gno

```diff
@@ -40,194 +38,222 @@ func rewardMathComputeInternalRewardAmount(tokenId uint64) uint64 {
 
 	lastCollectHeight := positionCollectHeight[tokenId]
 
-	avgGnsBlockAmount := gnsFullAmount / (uint64(currentHeight) - lastCollectHeight)
+	q96 := u256.MustFromDecimal(consts.Q96)
+	gnsFullAmountX96 := new(u256.Uint).Mul(u256.NewUint(gnsFullAmount), q96)
+	avgGnsBlockAmountX96 := new(u256.Uint).Div(gnsFullAmountX96, u256.NewUint(uint64(currentHeight)-lastCollectHeight))
+	println("avgGnsBlockAmountX96:", avgGnsBlockAmountX96.ToString())
+
+	gnsAmountToGiveX96 := u256.Zero()
 
 	// lastCollectHeight
 	// safeSub
 	switch durationRatio {
 	case 100:
-		if uint64(lastCalculatedHeight) > begin100 {
-			println("REWARD__REWARD ONLY 100%")
+		if uint64(lastCollectHeight) > begin100 {
```

Comment:
Memo for future refactoring task (Not sure is it working):

```go
// ...
switch durationRatio {
    case 100:
        gnsAmountToGive = calculateReward(currentHeight, lastCollectHeight, avgGnsBlockAmountX96, []struct {
            untilHeight uint64
            ratio       uint64
        }{
            {until100, 100},
            {until70, 70},
            {until50, 50},
            {until30, 30},
        })
    case 70:
        gnsAmountToGive = calculateReward(currentHeight, lastCollectHeight, avgGnsBlockAmountX96, []struct {
            untilHeight uint64
            ratio       uint64
        }{
            {until70, 70},
            {until50, 50},
            {until30, 30},
        })
    case 50:
        gnsAmountToGive = calculateReward(currentHeight, lastCollectHeight, avgGnsBlockAmountX96, []struct {
            untilHeight uint64
            ratio       uint64
        }{
            {until50, 50},
            {until30, 30},
        })
    case 30:
        gnsAmountToGive = calcAmount(avgGnsBlockAmountX96, currentHeight-lastCollectHeight, 30)
    default:
        // No reward
    }
    
// ...
    
func calculateReward(currentHeight, lastCollectHeight uint64, avgGnsBlockAmountX96 *uint256.Int, ratios []struct {
    untilHeight uint64
    ratio       uint64
}) uint64 {
    var gnsAmountToGive uint64
    lastHeight := lastCollectHeight

    for i, r := range ratios {
        if lastHeight >= r.untilHeight {
            continue
        }

        endHeight := currentHeight
        if i+1 < len(ratios) && endHeight > ratios[i+1].untilHeight {
            endHeight = ratios[i+1].untilHeight
        }

        duration := endHeight - lastHeight
        amount := calcAmount(avgGnsBlockAmountX96, duration, r.ratio)
        gnsAmountToGive += amount

        lastHeight = endHeight
        if lastHeight >= currentHeight {
            break
        }
    }

    return gnsAmountToGive
}
```

Reply:
tested but it's missing certain situtation